Implement quick and dirty error message solution plus a few additional accessibility fixes

bacall213 opened the pull request:
1. Implement the quick and dirty solution from #563 
2. Add a few minor CSS tweaks to keep text boxes from shuffling around when receiving and losing focus

bacall213 committed:
Update SettingsComponent.tsx

- Add minimum value (500ms) to toast duration input field
- Add "(opens in new window)" to links that open in new windows (https://www.w3.org/TR/WCAG20-TECHS/G201.html).
- Move a period out of the link one of the opens in a new window links.

bacall213 committed:
Fix text input box shuffle when gaining and losing focus

The border for focused boxes is 2px but the border for unfocused boxes is only 1px. By setting a 1px margin by default we create extra empty space and then change the margin back to zero when the box is focused so the 2px border takes up that space.

bacall213 committed:
Quick and dirty solution to fix a nonsensical error message

https://github.com/david-tejada/rango/issues/563

bacall213 committed:
Fix the shuffling text box issue for list select boxes

bacall213 committed:
Autocorrect toast duration to 500 (min) when invalid

Setting a min value in NumberInput only changes the limits for the input field but won't prevent someone from entering a number less than the minimum. This forces values to be at least the minimum (500ms).

bacall213 committed:
Remove some test statements

bacall213 commented:
I've been playing around with this a bit more and I have some additional improvements to add when it comes to contrast and focus. I also think I found a way to get the errors announced properly. Setting the alert div to role="alert" and aria-live="polite" is one part of it, and then passing in the box label to the error so the error message has context associated with it is the other part. The resulting readout to screen readers is is something like "Border radius (px) Please enter a value less than 72"

david-tejada commented:
I think for the external links we could use an icon and then have the text "Opens in new tab" as a .sr-only span after the icon. I would create a component to simplify this.

```tsx
import { faExternalLink } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

export function ExternalLink({
	href,
	children,
}: {
	readonly href: string;
	readonly children: React.ReactNode;
}) {
	return (
		<a href={href} target="_blank" rel="noreferrer">
			{children} <FontAwesomeIcon icon={faExternalLink} aria-hidden="true" />
			<span className="sr-only">Opens in new tab</span>
		</a>
	);
}
```

You need to create a .sr-only class in index.css:

```css
.sr-only {
	position: absolute;
	width: 1px;
	height: 1px;
	padding: 0;
	margin: -1px;
	overflow: hidden;
	clip: rect(0, 0, 0, 0);
	white-space: nowrap;
	border-width: 0;
}
```

Then we can use it like this:

```tsx
<p className="small show-on-focus">
  Use a{" "}
  <ExternalLink href="https://developer.mozilla.org/en-US/docs/Web/CSS/color_value">
    CSS color string
  </ExternalLink>
  .
</p>
```

As for the other changes, you can add them to this PR since it's all related to accessibility or create a new one. Whatever you prefer.

bacall213 committed:
Update SettingsComponent.tsx

Replace open in new tab implementation with David's suggested solution for external links

bacall213 committed:
Update ExcludeKeysSetting.tsx

1. Move focus back to the add button after deleting an exclude keys entry. 

Focus is currently being returned to an invisible spot  instead of the actual button. This change returns it to the add button every time. 

If we return it to the delete button, if a row exists, then a user who clicks twice accidentally might delete an entry they wanted. Moving focus back to the Add button helps protect the user from making a change they can't undo.

useRef was added for this and then called in the handleDeletion() function: `addExcludeKeyButtonRef.current?.focus();`

2. Move the explanation text for exclude keys up so the explanation comes first.

3. Change the icon colors to darker colors that have better contrast for a white or black background. Also change the icon color references to use the variables used elsewhere in settings so they'll remain consistent.

4. Change the icons to the larger version. For the trash can, in particular, this makes the target more visible.

5. Add aria label "Add exclude keys pattern" to the add button so it's clear for screen reader users what the button does.

6. Add aria label to the keys to exclude text field so the purpose of the field is more apparent.

7. Add an aria label to the trashcan that makes it apparent what pattern will be deleted: `aria-label={"Delete pattern " + entry[0]}`

bacall213 committed:
Update CustomHintsSetting.tsx

1. Use color variable references for the icons to keep things consistent.

2. Bump the size of some of the icons so the target is easier to hit.

3. Use `useRef` to assign a reference to the "Add selector to include" button so that we can return focus back to that button after a row is deleted. Similar to exclude keys, this is the safer option since there is no undo for deleting an entry.

bacall213 committed:
Update index.css

1. Add an in-between slate color (slate-350) to use for the toggle backgrounds so a disabled toggle maintains at least 4.5:1 contrast ratio regardless of whether the page has a white or black background (in case you implement dark mode later). Since the toggles are relatively small and they're control elements, I choose a color that works with the WCAG AA text requirement instead of just the graphical element requirement.

2. Change green-500 and red-500 to darker variants that are more contrasting (meeting >= 4.5:1). The trashcan doesn't have any text label associated with it so the higher contrast is needed. For the toggles, the green will maintain >= 4.5:1 contrast for both a white and black background.

3. Add the screen reader only formatting for the "open in new tab" addition.

4. Change the button size to 1rem so the buttons match with the surrounding elements.

bacall213 committed:
Update NumberInput.tsx

Wrap the alert text in a div so there's a spot in the DOM for the alert to appear instead of the entire thing needing to be created. It might be possible to make it work without this, but when I was testing the alerts with VoiceOver it seemed like I needed something already present that the aria-live component drops into for it to work correctly.

bacall213 committed:
Update toggleHints.ts

Call notifyTogglesStatus() after toggling the hints at a global level. This results in a status toast whether you click the icon or use the keyboard shortcut. For reasons I couldn't track down, the status toast was only being displayed when called with Talon directly.

Displaying the status when the icon is clicked or activated will help users understand what happened if they perform the action accidentally.

bacall213 committed:
Update Toast.css

Use a darker base color in the toast notifications.

bacall213 committed:
Update toastify.css

Additional color adjustments for toast notifications. This also makes the close button more contrasty.

bacall213 committed:
Update ToastTogglesMessage.css

Increase the contrast of the "unset" indicators.

bacall213 committed:
Update ToastIcon.tsx

Minor color adjustments to improve contrast ratios. The green matches green-500 and the red matches red-500.

bacall213 committed:
Update Alert.tsx

1. Pass the element ID into the alert so we can attach it to the same label as the input box that generated the error. 

2. Assign the aria role="alert" and aria-live="polite" to announce the errors (politely, of course)

bacall213 commented:
One of the lingering issues is that the toast notifications use an aria-hidden icon for reporting the state. This is mostly an issue for the list of hint statuses. The toast is set to role="alert" by default so it's being read by screen readers, but the actual useful values for the states aren't being read. I found a way to add aria-labels to them and reference the state variable, but since the toast doesn't receive focus (and I'm not sure that it should) the aria-labels aren't read. 

I'm not sure yet what the best way to solve this one is.

bacall213 committed:
Update onboarding page

Change link text so the purpose of the link is more clear.

bacall213 committed:
Add main landmark and move footer in what's new

Add a "main" landmark to the overall div and then move the footer out of it so there aren't nested landmarks. Adding a <main> tag wrapping <div class="container"> would have worked as well.

bacall213 committed:
Add main landmark to onboarding page

bacall213 commented:
The toast notifications look like they would be considered a [status message](https://www.w3.org/WAI/WCAG21/Understanding/status-messages.html) and shouldn't receive focus, but the underlying content still needs to be announced. Still trying to find a way to solve for that. 

After some more testing, the alert boxes don't need the extra div tags to work reliably so I'll remove them from the updated file in the pull request.

david-tejada commented on src/pages/settings/SettingsComponent.tsx:
```diff
@@ -18,6 +20,21 @@ let justSaved = false;
+		</a>
+	);
+}
+
```
Better move this to its own file

bacall213 committed:
Update Alert.tsx

- Assign the aria role based on the type passed into the alert.
- Cleanup some of the extra pieces that didn't work as well.

bacall213 committed:
Create ExternalLink.tsx

Move ExternalLink function into separate file per comment on pull request.

bacall213 committed:
Update TextInput.tsx

- Add updated validation bits

bacall213 committed:
Update NumberInput.tsx

- Update validation bits using the zod validation features in the settings schema

bacall213 committed:
Update SettingsComponent.tsx

- Move ExternalLink bits out to separate file and import it
- Add updated validation pieces

Note on validation - I couldn't figure out which type was needed to handle an object and then reliably get the boolean for 'valid' out of it and the optional 'message' field without typescript throwing errors so I kept them separate as 'isValid' (still a boolean) and 'validationMessage' (string)

bacall213 committed:
Update ExcludeKeysSetting.tsx

Additional changes to the button labeling for screen reader users.

bacall213 committed:
Update CustomHintsSetting.tsx

Additional changes to button labeling for screen reader users.

bacall213 committed:
Update settings.ts

Add the checkValidity function David mentioned earlier in the pull request discussion.

bacall213 committed:
Update settingsSchema.ts

Update the settings schema with appropriate validation error messages for various fields.

bacall213 committed:
Make sr-only CSS class available for toast notifications

bacall213 committed:
Add additional context for screen reader users

Changes the title of the notifications to "Rango notifier:" for screen reader users. When screen readers read out the toast notification, the semantic relationships are lost and all of the text runs together. For the toggle status notification, the result is something like "Rango Now Page Host Global dismiss to close"

This change, combined with the others, help make the readout make more sense for screen reader users without affecting the visual formatting.

bacall213 committed:
Add screen reader enhancements

Utilize the sr-only css class to add additional content to the readout of notifications for screen reader users.

I initially tried to solve for this by adding labels to the icons and unhiding them from screen readers, but the toggle status notification puts the icons on the right and the label on the left whereas the other notifications put the icon on the left and the text on the right. The non-toggle notifications already included the status information in the accessible text. Adding the label to the image made them nonsensical.

Instead, this adds additional helper text after the visual label that's only exposed to screen reader users. getIconType() returns the exact text needed, so I used that in the screen reader output.

The expected result for a toggle status dialog is something like: "Rango notifier: Now hints are unset. Page hints are unset. host hints are disabled. Global hints are enabled. dismiss to close"

For non-toggle notifications, the result is something like "Rango notifier: keyboard clicking enabled"

bacall213 committed:
Add additional row position details when deleting an exclude keys row

This provides additional positional information for screen reader users since the rows are not presented using standard table elements.

bacall213 committed:
Add additional positional hints when deleting a custom hints row

Specific to screen reader users since the content isn't rendered using table elements. There are no visible changes that will result from this.

bacall213 committed:
Move color string validation messages into settings schema

bacall213 committed:
Move color string validation message out of TextInput

Moved to settingsSchema

bacall213 committed:
Change the background color alert into an explanation

The alert dialog was a bit too much. This follows the same pattern in the other components where an explanation message is made visible when a field is disabled.

bacall213 committed:
Add additional formatting for validation alerts for custom hints

bacall213 committed:
Add custom hint validation messaging that isn't color-based

The regex and selector fields both performed validation, but didn't reset the value when the user leaves the field even if the value didn't parse correctly. The validation styling relied on color alone to alert the user to an issue.

This adds a separate block to display validation alerts for the custom hints.

I wanted to put the alerts immediately below the appropriate box or row, but I couldn't get it to work properly in a way that passed the typescript tests. If you know of a better way to add on the alerts, I'm open to alternatives.

This change also adds aria-invalid to the regex text fields, but not to the selector text fields. When adding it to the selector text field, it technically works (and works well), but it gets caught as having an invalid value. 

`aria-invalid={selector && !isValidSelector(selector)}` is what didn't work.

bacall213 committed:
Remove the extra validation message logic (moved to schema)

bacall213 committed:
Cleanup unneeded isValid() function

bacall213 committed:
Improve error message handling in schema

A 'message' can be defined for some (all?) zod objects, including z.number(). By assigning an error message to the 'message' property for each schema item (where relevant), we negate the need for the extra logic in NumberInput.tsx to handle NaN instances. The value of 'message' will be returned instead. When the value is outside of the .refine() range, invalid input is returned. We can also apply the same message to that.

Instead writing the same message twice, errors = {...} was added at the top to store the various error messages.

There might be a better way to do this with a custom error map in zod, but this seems to work quite well.

bacall213 committed:
Adjust styling for new div

bacall213 committed:
More refined error output

The previous push added a second iterator which was probably not resource friendly. To get around that, I had to wrap the entire block in another div which then required some additional CSS changes to keep the formatting consistent. 

One benefit of this change is that now the alert boxes will appear immediately below the row where the error is occuring.

david-tejada committed:
Fix autofocus when adding keys to exclude from context menu

david-tejada commented:
I added a commit. It has nothing to do with your changes. I just noticed it and since it's related I added it here.

What's the status of this PR, are you going to make more changes or is it ready to merge? I want to make a release soon and I want to include it.

bacall213 commented:
Everything is stable in it, but give me another day to double-check if there's anything else that needs to be added and also to tune the regex and selector error box formatting so they're consistently sized while being flexible. That will help ensure they look like they were there from the start.

bacall213 committed:
Update Toggle.css

- Add CSS for prefers reduced motion to disable toggle animations if the user doesn't want animations.
- Use system color GrayText for the background for disabled toggles so it's more immediately apparent that the toggles are in a different state.
- Make the toggles a hair larger so that the toggles still look correct in high contrast mode when the text size for buttons is set to 1rem.

bacall213 committed:
Add borders for high contrast mode

- Adds a border around the toasts when the browser is set to high contrast mode.
- The background styling for <code> text is lost in high contrast, so add back a thin border to differentiate it.

bacall213 committed:
Add border to alerts in high contrast mode

bacall213 committed:
Add borders for high contrast mode

bacall213 committed:
Add a thin border to <code> elements when using high contrast mode

bacall213 committed:
Update CustomHintsSetting.css

Update the presentation of the alert boxes so they take up the full width. It looks a bit better than the expanding box.

bacall213 committed:
Reduce min-width for wrapping at high zooms

Reduce the min-width so the content wraps for more people using browser zoom at 200%+ and/or lower resolutions without the need for horizontal scrolling.

bacall213 committed:
Change "Add" button to "Add keys to exclude"

I added an aria-label for the 'Add' button, but since there was visible text the aria label and visible text really should be the same, so I changed the button label to "Add keys to exclude" (to be consistent with the other wording on the page) and then removed the aria label since the there was no longer a need for it.

david-tejada commented:
Sure, no problem. Just let me know when it's ready

bacall213 committed:
Add break-all to <code> text in what's new

When the viewport is particularly small, there are blocks of text that typically won't wrap naturally. This adds a horizontal scroll bar. Adding break-all to the <code> blocks forces a wrap for the very few (maybe one) cases where it's a problem. All other wrapping defaults to word-based wrapping.

bacall213 commented:
Alright, that should be good. Do you want any comparison screenshots for how the PR changes things? Or are you already checking on your end?

david-tejada commented:
No need for that. I will check myself. Thanks for all the work!

david-tejada committed:
Remove unnecessary error handling of deprecated setting hintMinimumContrastRatio
